xtensa: Add initial support for ESP32-S3

gustavonihei opened the pull request:
## Summary
This PR intends to add basic support for **ESP32-S3**, a dual-core Xtensa LX7 MCU from Espressif Systems.\
https://www.espressif.com/en/products/socs/esp32-s3

The new `esp32s3-devkit` target considers a generic development board whose characteristics are common to [ESP32-S3-DevKitC-1](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html) and [ESP32-S3-DevKitM-1](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/hw-reference/esp32s3/user-guide-devkitm-1.html) official development kits.

This initial support consists of the following features:
- Basic clock control (APB clock defaults to 80MHz)
- Peripheral and CPU interrupts management
- Heap management
- Board reset support
- Serial driver

## Impact
Support for new chip.

## Testing
`esp32s3-devkit:nsh` and successful execution of `ostest` application.

gustavonihei commented:
Please, ignore the Check failure, the `tie-asm.h` and `tie.h` files come from the Xtensa Overlays. We'd better include them unchanged.

pkarashchenko commented on arch/xtensa/src/esp32s3/.gitignore:
```diff
@@ -0,0 +1 @@
+/esp-nuttx-bootloader
```
```suggestion
/esp-nuttx-bootloader

```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.c:
```diff
@@ -0,0 +1,851 @@
+
+  if (length >= UART_DATA_5_BITS && length <= UART_DATA_8_BITS)
+    {
+      modifyreg32(UART_CONF0_REG(priv->id), UART_BIT_NUM_M,
+                    length << UART_BIT_NUM_S);
```
```suggestion
      modifyreg32(UART_CONF0_REG(priv->id), UART_BIT_NUM_M,
                  length << UART_BIT_NUM_S);
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.c:
```diff
@@ -0,0 +1,851 @@
+int esp32s3_lowputc_data_length(const struct esp32s3_uart_s * priv)
+{
+  int ret = OK;
+  uint32_t length = (priv->bits - 5);
```
```suggestion
  uint32_t length = priv->bits - 5;
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.c:
```diff
@@ -0,0 +1,851 @@
+uint32_t esp32s3_lowputc_get_sclk(const struct esp32s3_uart_s * priv)
+{
+  uint32_t clk_conf_reg;
+  uint32_t ret = -ENODATA;
+  clk_conf_reg   = getreg32(UART_CLK_CONF_REG(priv->id));
+  clk_conf_reg  &= UART_SCLK_SEL_M;
+  clk_conf_reg >>= UART_SCLK_SEL_S;
```
```suggestion
  uint32_t ret   = -ENODATA;
  clk_conf_reg   = getreg32(UART_CLK_CONF_REG(priv->id));
  clk_conf_reg  &= UART_SCLK_SEL_M;
  clk_conf_reg >>= UART_SCLK_SEL_S;
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.c:
```diff
@@ -0,0 +1,851 @@
+  reg = getreg32(UART_STATUS_REG(priv->id));
+  reg = reg >> UART_TXFIFO_CNT_S;
+  reg = reg & UART_TXFIFO_CNT_V;
+  if (reg < (UART_TX_FIFO_SIZE -1))
+    {
+      return false;
+    }
+  else
+    {
+      return true;
+    }
```
```suggestion
  return !(reg < (UART_TX_FIFO_SIZE - 1));
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.h:
```diff
@@ -0,0 +1,485 @@
+
+enum uart_stop_length
+{
+    UART_STOP_BITS_1   = 0x1,  /* Stop bit: 1 bit */
+    UART_STOP_BITS_2   = 0x3,  /* Stop bit: 2 bits */
```
```suggestion
  UART_STOP_BITS_1   = 0x1,  /* Stop bit: 1 bit */
  UART_STOP_BITS_2   = 0x3,  /* Stop bit: 2 bits */
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+
+#ifdef CONSOLE_UART
+#  if defined(CONFIG_UART0_SERIAL_CONSOLE)
+#    define CONSOLE_DEV     g_uart0_dev     /* UART0 is console */
+#    define TTYS0_DEV       g_uart0_dev     /* UART0 is ttyS0 */
+#    define UART0_ASSIGNED      1
```
```suggestion
#    define CONSOLE_DEV         g_uart0_dev     /* UART0 is console */
#    define TTYS0_DEV           g_uart0_dev     /* UART0 is ttyS0 */
#    define UART0_ASSIGNED      1
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+
+  /* Tx fifo empty interrupt or UART tx done int */
+
+  if (int_status & tx_mask)
```
```suggestion
  if ((int_status & tx_mask) != 0)
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+
+  /* Rx fifo timeout interrupt or rx fifo full interrupt */
+
+  if (int_status & rx_mask)
```
```suggestion
  if ((int_status & rx_mask) != 0)
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+  status_reg = getreg32(UART_STATUS_REG(priv->id));
+  bytes = status_reg & UART_RXFIFO_CNT_M;
+
+  return (bytes > 0) ? true : false;
```
```suggestion
  return (bytes > 0);
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+
+static bool esp32s3_txready(struct uart_dev_s *dev)
+{
+  return (esp32s3_lowputc_is_tx_fifo_full(dev->priv)) ? false : true;
```
```suggestion
  return !esp32s3_lowputc_is_tx_fifo_full(dev->priv);
```

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+  reg = getreg32(UART_INT_RAW_REG(priv->id));
+  reg = reg & UART_TXFIFO_EMPTY_INT_RAW_M;
+
+  return (reg > 0) ? true : false;
```
```suggestion
  return (reg > 0);
```

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_cache_memory.h:
```diff
@@ -0,0 +1,118 @@
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H_
+#define __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H_
```
```suggestion
#ifndef __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H
#define __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H
```

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_cache_memory.h:
```diff
@@ -0,0 +1,118 @@
+#define CACHE_MEMORY_DBANK0_ADDR        0x3fcf0000
+#define CACHE_MEMORY_DBANK1_ADDR        0x3fcf8000
+
+#endif /* __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H_ */
```
```suggestion
#endif /* __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_CACHE_MEMORY_H */
```

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_rtccntl.h:
```diff
@@ -0,0 +1,5795 @@
+ ****************************************************************************/
+
+#ifndef __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTC_CNTL_H
+#define __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTC_CNTL_H
```
```suggestion
#ifndef __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTCCNTL_H
#define __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTCCNTL_H
```

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_rtccntl.h:
```diff
@@ -0,0 +1,5795 @@
+#define RTC_CNTL_DATE_V  0x0fffffff
+#define RTC_CNTL_DATE_S  0
+
+#endif /* __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTC_CNTL_H */
```
```suggestion
#endif /* __ARCH_XTENSA_SRC_ESP32S3_HARDWARE_ESP32S3_RTCCNTL_H */
```

pkarashchenko reviewed (commented):
Do we need `boards/xtensa/esp32s3/.gitignore`? It is empty

Ouss4 commented on arch/xtensa/include/esp32s3/core-isa.h:
```diff
@@ -0,0 +1,695 @@
+/****************************************************************************
+ * arch/xtensa/include/esp32s3/core-isa.h
+ * Xtensa processor core configuration information.
+ *
+ * Customer ID=15128; Build=0x90f1f; Copyright (c) 1999-2021 Tensilica Inc.
```
Yes, we need to keep the original header here.  In all of the Xtensa folder we have some files that come from Tensilica and we can't change the headers to the ASF one.

Ouss4 commented on arch/xtensa/src/esp32s3/esp32s3_clockconfig.c:
```diff
@@ -0,0 +1,312 @@
+ *
+ ****************************************************************************/
+
+extern void ets_update_cpu_frequency(uint32_t ticks_per_us);
```
These functions come from the ROM ELF file.  The prototypes are just declared to avoid warnings.

pkarashchenko replied:
Is there an option to create `esp32s3_romapi.h` to list all needed APIs or this is too big effort?

Ouss4 replied:
The API is defined in the linker scripts at `boards/xtensa/esp32s3/esp32s3-devkit/scripts/esp32s3_rom.ld` (they are in the `common` directory for ESP32.).  The thing is there are a lot of functions there and we just use a few of them, so each source file just adds the prototypes of those it uses.

Ouss4 commented on arch/xtensa/src/esp32s3/esp32s3_lowputc.c:
```diff
@@ -0,0 +1,851 @@
+{
+  .periph = ESP32S3_PERIPH_UART0,
+  .id = 0,
+  .cpuint = -ENOMEM,
```
There are situations where `cpuint` is tested to verify the validity of an operation, any error code would work for this however, the function that allocates a CPU interrupt returns `ENOMEM` in case of failure, "Out of Memory" here means "Out of CPU Interrupts".

pkarashchenko replied:
Thank you for explanation

Ouss4 commented on arch/xtensa/src/esp32s3/esp32s3_irq.c:
```diff
@@ -0,0 +1,686 @@
+   * available.
+   */
+
+  cpu = up_cpu_index();
```
ESP32-S3 is a dual core chip, so yes we do expect an index other than `0`.  However, for this initial support all SMP features are now ignored.  These will be dealt with later.
Regarding `volatile uint32_t *g_current_regs[1];` this will change to the usual:
```
#ifdef CONFIG_SMP
/* For the case of architectures with multiple CPUs, then there must be one
 * such value for each processor that can receive an interrupt.
 */

volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];

#else

volatile uint32_t *g_current_regs[1];

#endif
```

pkarashchenko replied:
Recently we had https://github.com/apache/incubator-nuttx/pull/5264 so we can go with `volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];` even right now. But this is minor comment. It is fine to go with `volatile uint32_t *g_current_regs[1];` for now and rework later

gustavonihei commented:
> Do we need `boards/xtensa/esp32s3/.gitignore`? It is empty

No need, added for creating the initial tree and forgot to remove it later. Thanks.

gustavonihei commented on arch/xtensa/include/esp32s3/tie.h:
```diff
@@ -0,0 +1,210 @@
+
+#define XCHAL_CP_NUM			2	/* number of coprocessors */
+#define XCHAL_CP_MAX			4	/* max CP ID + 1 (0 if none) */
+#define XCHAL_CP_MASK			0x09	/* bitmask of all CPs by ID */
```
Done.

gustavonihei commented:
@pkarashchenko I've updated the PR addressing your comments.

Also, I've moved the change to the Dockerfile to another PR. If that gets merged first, then the `esp32s3-devkit:nsh` may be built by the CI on this PR.

pkarashchenko commented on arch/xtensa/src/esp32s3/esp32s3_timerisr.c:
```diff
@@ -0,0 +1,143 @@
+
+  /* Attach the timer interrupt. */
+
+  irq_attach(ESP32S3_IRQ_SYSTIMER_TARGET0, (xcpt_t)systimer_isr, NULL);
```
Maybe we can remove `(xcpt_t)` so compiler will generate warning in case if something is wrong?

gustavonihei commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+    case TIOCSERGSTRUCT:
+      {
+         struct esp32s3_uart_s *user = (struct esp32s3_uart_s *)arg;
+         if (!user)
```
Actually the correct would be `user == NULL`. Thanks!

gustavonihei commented on arch/xtensa/src/esp32s3/esp32s3_allocateheap.c:
```diff
@@ -0,0 +1,89 @@
+   */
+
+  extern uint8_t *_sheap;
+  extern const struct esp32s3_rom_layout_s *ets_rom_layout_p;
```
I've removed these declarations from here.
`_sheap` is already declared in `xtensa.h` and `ets_rom_layout_p` has been moved to the `esp32s3_rom_layout.h` header.

gustavonihei commented on arch/xtensa/src/esp32s3/esp32s3_serial.c:
```diff
@@ -0,0 +1,1166 @@
+  /* Disassociate the peripheral interrupt from the CPU interrupt */
+
+  esp32s3_teardown_irq(priv->cpu, priv->periph, priv->cpuint);
+  priv->cpuint = -1;
```
Done.

pkarashchenko commented on boards/xtensa/esp32s3/esp32s3-devkit/scripts/esp32s3.ld:
```diff
@@ -0,0 +1,269 @@
+    *(.gnu.version_r)
+    *(.eh_frame)
+
+    . = (. + 3) & ~ 3;
```
Maybe just `. = ALIGN(4);`?

gustavonihei replied:
Sure, much better this way. Thanks!

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_systimer.h:
```diff
@@ -0,0 +1,808 @@
+ * register file clk gating
+ */
+
+#define SYSTIMER_CLK_EN    (BIT(31))
```
minor
```suggestion
#define SYSTIMER_CLK_EN    BIT(31)
```
here and in other places

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_system.h:
```diff
@@ -0,0 +1,1757 @@
+ * Set 1 to let core1 reset
+ */
+
+#define SYSTEM_CONTROL_CORE_1_RESETING    (BIT(2))
```
minor
```suggestion
#define SYSTEM_CONTROL_CORE_1_RESETING    BIT(2)
```
here and in other places

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_rtccntl.h:
```diff
@@ -0,0 +1,5795 @@
+ * SW system reset
+ */
+
+#define RTC_CNTL_SW_SYS_RST    (BIT(31))
```
minor
```suggestion
#define RTC_CNTL_SW_SYS_RST    BIT(31)
```
here and in other places

pkarashchenko commented on arch/xtensa/src/esp32s3/hardware/esp32s3_iomux.h:
```diff
@@ -0,0 +1,467 @@
+
+/* Output enable in sleep mode */
+
+#define SLP_OE     (BIT(0))
```
minor
```suggestion
#define SLP_OE     BIT(0)
```
here and in other places

pkarashchenko reviewed (commented):
Only few cosmetic changes if possible. Other looks good!

gustavonihei commented on arch/xtensa/src/esp32s3/hardware/esp32s3_uart.h:
```diff
@@ -0,0 +1,1961 @@
+ * mode.
+ */
+
+#define UART_WAKEUP_INT_RAW    (BIT(19))
```
@pkarashchenko These are auto-generated, so I prefer to leave these as-is to avoid additional work for every other future header.

pkarashchenko commented:
LGTM!
